src/QsCompiler/CompilerOptimizations/StatementRemoving.fs

```diff
@@ -0,0 +1,129 @@
+module Microsoft.Quantum.QsCompiler.CompilerOptimization.StatementRemoving
+
+open System
+open System.Collections.Immutable
+open Microsoft.Quantum.QsCompiler.SyntaxExtensions
+open Microsoft.Quantum.QsCompiler.SyntaxTree
+open Microsoft.Quantum.QsCompiler.Transformations.Core
+
+open ComputationExpressions
+open Utils
+open OptimizingTransformation
+open VariableRenaming
+
+
+/// Returns whether all variables in a symbol tuple are discarded
+let rec private isAllDiscarded = function
+| DiscardedItem -> true
+| VariableNameTuple items -> Seq.forall isAllDiscarded items
+| _ -> false
+
+/// Returns whether a statements is purely classical.
+/// The statement must have no classical or quantum side effects other than defining a variable.
+let private isPureClassical = function
+| QsVariableDeclaration s when not (s.Rhs.InferredInformation.HasLocalQuantumDependency) -> true
+| _ -> false
+
+/// Returns whether a statement is purely quantum.
+/// The statement must have no classical side effects, but can have quantum side effects.
+let private isPureQuantum = function
+| QsVariableDeclaration s when isAllDiscarded s.Lhs -> true
+| QsExpressionStatement _ -> true
+| _ -> false
+
+/// Reorders a list of statements such that the pure classical statements occur before the pure quantum statements
+let rec private reorderStatements (stmtList: QsStatement list): QsStatement list =
+    match stmtList with
+    | [] -> []
+    | head :: tail ->
+        let tail = reorderStatements tail
+        match head :: tail with
+        | head1 :: head2 :: tail
+            when isPureQuantum head1.Statement && isPureClassical head2.Statement ->
+            head2 :: reorderStatements (head1 :: tail)
+        | a -> a
+
+
+/// The SyntaxTreeTransformation used to remove useless statements
+type internal StatementRemover() =
+    inherit OptimizingTransformation()
+
+    /// The VariableRenamed used to ensure unique variable names
+    let mutable renamer = None
+
+    /// Returns whether the given statement is needed.
+    /// If this returns false, the given statement can be removed without changing the code.
+    let isStatementNeeded stmt =
+        match stmt.Statement with
+        | QsVariableDeclaration s when isAllDiscarded s.Lhs && not (s.Rhs.InferredInformation.HasLocalQuantumDependency) -> false
+        | QsScopeStatement s when s.Body.Statements.IsEmpty -> false
+        | QsForStatement s when s.Body.Statements.IsEmpty -> false
+        | QsWhileStatement s when s.Body.Statements.IsEmpty -> false
```

Comment:
Please verify that the condition does not have a quantum dependency such that you are not relying on the assumption that the condition cannot have a quantum dependency. 

Also, an if statement is not needed if all blocks are empty and none of the conditions have any quantum dependencies. Similarly for a repeat statement, and in particular also for a qubit scope.